train_utils.py

```diff
@@ -55,18 +179,35 @@ def fit(epochs,
         train_data:
             A list of tuples of positive pair of questions (question_text1, question_text2)
 
-        val_data:
+        test_data:
             A list of tuples of positive pair of questions (question_text1, question_text2)
+
+        candidates:
+            A list of tuples of (id, text)
+
+        top_k:
+            Use top K candidates for computing mean average precision
+
     """
     for epoch in range(epochs):
+        model.train()
+
         for q1_batch, q2_batch in iterate_batch(train_data, batch_size):
-            loss_batch(model, loss_func, q1_batch, q2_batch, opt)
+            loss_batch(model, loss_func, q1_batch, q2_batch, pairwise_similarity_func, opt)
 
         model.eval()
         with torch.no_grad():
             losses, nums = zip(
-                *[loss_batch(model, loss_func, xb, yb) for xb, yb in iterate_batch(val_data, batch_size)]
+                *[loss_batch(model, loss_func, xb, yb, pairwise_similarity_func)
+                  for xb, yb in iterate_batch(test_data, batch_size)]
```

Comment:
Nit: I think this should be called validation_data. 

> Training Dataset: The sample of data used to fit the model.
> Validation Dataset: The sample of data used to provide an unbiased evaluation of a model fit on the training dataset while tuning model hyperparameters. The evaluation becomes more biased as skill on the validation dataset is incorporated into the model configuration.
> Test Dataset: The sample of data used to provide an unbiased evaluation of a final model fit on the training dataset.

https://machinelearningmastery.com/difference-test-validation-datasets/

Reply:
Also, does it make sense to split validation data into mini batches vs evaluating on all the data at once?

Reply:
To answer my own question, it does:
> In-batch precision@1 (accuracy computed over each row of the similarity matrix M)